### PR TITLE
implemented open_popup and close_popup methods

### PR DESCRIPTION
**Pull request by PROgram52bc**
#912

### CODE REVIEW

File: ipyleaflet/leaflet.py [hunk 1]
```diff
@@ -497,6 +497,11 @@ class Popup(UILayer):
     auto_close = Bool(True).tag(sync=True, o=True)
     close_on_escape_key = Bool(True).tag(sync=True, o=True)
 
+    def open_popup(self, location=def_loc):
+        self.send({'msg': 'open', 'location': location})
```
Comment (martinRenou) [inline ipyleaflet/leaflet.py:501]: There seems to be a duplication of information between the `location` property and the `location` argument of this method.

Should this method default to the location property?
```python
def open_popup(self, location=None):
    self.send({'msg': 'open', 'location': location is None ? self.location : location})
```

And should we change the location property if it's provided?
```python
def open_popup(self, location=None):
    if location is not None:
        self.location = location
    self.send({'msg': 'open', 'location': location is None ? self.location : location})
```

Comment (martinRenou) [reply ipyleaflet/leaflet.py:501]: I realize I've been writing non-valid Python code in my comment ahah. Let's try again:

```python
def open_popup(self, location=None):
    self.send({'msg': 'open', 'location': self.location if location is None else location})
```

And should we change the location property if it's provided?
```python
def open_popup(self, location=None):
    if location is not None:
        self.location = location
    self.send({'msg': 'open', 'location': self.location if location is None else location})
```

**Comment by martinRenou** [comment]
flake8 is complaining in the CI test:
```
ipyleaflet/leaflet.py:506:1: E302 expected 2 blank lines, found 1
```

I think we can ignore the UI tests failure, it looks unrelated.

**Comment by PROgram52bc** [comment]
> flake8 is complaining in the CI test:
> 
> ```
> ipyleaflet/leaflet.py:506:1: E302 expected 2 blank lines, found 1
> ```
> 
> I think we can ignore the UI tests failure, it looks unrelated.

Thanks! There is another failure in the Visual Regression test. Do you have any idea what is causing that test to fail?

**Comment by martinRenou** [comment]
> Thanks! There is another failure in the Visual Regression test. Do you have any idea what is causing that test to fail?

It looks like the basemap has changed a bit. You can ignore that. I don't think it's related to your change.

**Comment by PROgram52bc** [comment]
Do you think it is a good idea to also have an `is_open` attribute on the `Popup` class as mentioned in #239? What are some possible ways to implement that?

**Comment by martinRenou** [comment]
You can probably listen to the `popupopen`, and `popupclose` events on the Map https://leafletjs.com/reference-1.3.4.html#map-popupopen

But then you will need to be able to map those events to the right popup widgets?

File: ipyleaflet/leaflet.py [hunk 2]
```diff
@@ -497,6 +497,23 @@ class Popup(UILayer):
...
+        """
+
+        if location is not None:
+            self.location = location
+        self.send({'msg': 'open', 'location': self.location if location is None else location})
+
+    def close_popup(self):
+        """ Close the popup on the bound map """
+
+        self.send({'msg': 'close'})
```
Comment (martinRenou) [inline ipyleaflet/leaflet.py:517]: You need two spaces here for flake8 to be happy

Comment (PROgram52bc) [reply ipyleaflet/leaflet.py:517]: Two spaces where?

**Comment by PROgram52bc** [comment]
> But then you will need to be able to map those events to the right popup widgets?

I guess there is a reference to the popup object in the [popup event](https://leafletjs.com/reference.html#popupevent-popup)?

**Comment by martinRenou** [comment]
> I guess there is a reference to the popup object in the popup event?

To the popup object indeed, but there should not be a reference to the popup **widget**

File: ipyleaflet/leaflet.py (hunk 2 above)
Comment (martinRenou) [reply ipyleaflet/leaflet.py:517]: Two blank lines sorry. Between the class definitions.

**Comment by PROgram52bc** [comment]
> > I guess there is a reference to the popup object in the popup event?
> 
> To the popup object indeed, but there should not be a reference to the popup **widget**

I see. I guess we can tackle that issue in a separate PR then.

Any more comments?

**Comment by PROgram52bc** [comment]
> Sorry I wasn't clearer about the flake8 issue. It is still not resolved. I will push a commit to your branch fixing that if you don't mind.

Yeah sure, please feel free :)

**Comment by martinRenou** [comment]
Actually you fixed it! Sorry I was confused by the diff.

Let's wait for the CI to be green then I'll merge :)

Thanks a lot!

**Pull request merged**